### PR TITLE
tso: fix possible stale ts when leader changes twice.

### PR DESCRIPTION
**Pull request by disksing**
PTAL @siddontang @nolouch

### CODE REVIEW

File: server/leader.go [hunk 1]
```diff
@@ -191,6 +185,18 @@ func (s *Server) campaignLeader() error {
...
 	}
 
+	s.enableLeader(true)
+	defer func() {
+		s.enableLeader(false)
+		// Reset connections and cluster.
+		s.closeAllConnections()
+		s.cluster.stop()
+		// Clear ts.
+		s.ts.Store(&atomicObject{
```
Comment (siddontang) [inline server/leader.go]: seem you can set nil directly? so `s.ts.Load().(*atomicObject)` will return not ok.

File: server/leader.go [hunk 2]
```diff
@@ -177,8 +173,6 @@ func (s *Server) campaignLeader() error {
 	}
 
 	log.Debugf("campaign leader ok %s", s.Name())
-	s.enableLeader(true)
-	defer s.enableLeader(false)
 
 	// Try to create raft cluster.
 	err = s.createRaftCluster()
```
Comment (siddontang) [inline server/leader.go]: we should defer 
```
s.closeAllConnections()
s.cluster.stop()
```
after create the cluster successfully, before sync TSO.

File: server/leader.go (hunk 1 above)
Comment (disksing) [inline server/leader.go]: It will panic: https://github.com/golang/go/blob/master/src/sync/atomic/value.go#L50
Maybe we can `s.ts = atomic.Value{}`.

Comment (siddontang) [inline server/leader.go]: got it.

**Comment by nolouch** [comment]
LGTM

**Comment by ngaut** [comment]
Please add tests.

**Comment by disksing** [comment]
@ngaut I found it hard to construct a test case correspond to the issue. Basically I can't manually transfer embed EtcdServer's leadership because the method is not exported. See https://github.com/coreos/etcd/blob/master/etcdserver/server.go#L932
Or maybe we can test it by other ways? cc @siddontang

**Comment by siddontang** [comment]
Hi @disksing 

We don't need to transfer etcd leadership, the PD leader is not the same as etcd's. We can force update the leader key with another value, and then update with the old leader again.

**Comment by disksing** [comment]
@siddontang PTAL. After my test, just delete the leader key and wait for a while is enough, the servers will start campaign after they are notified by etcd.

File: pd-client/leader_change_test.go [hunk 3]
```diff
@@ -111,6 +120,49 @@ func (s *testLeaderChangeSuite) TestLeaderChange(c *C) {
 	c.Error("failed getTS from new leader after 10 seconds")
 }
 
+func (s *testLeaderChangeSuite) TestLeaderTransfer(c *C) {
+	servers, endpoints, closeFunc := s.prepareClusterN(c, 3)
```
Comment (siddontang) [inline pd-client/leader_change_test.go]: seem we can use 2 server here. So if we delete the leader, another can become leader directly, then we delete the new leader, the previous old leader may become leader again.

**Comment by siddontang** [comment]
LGTM

PTAL @overvenus

File: pd-client/leader_change_test.go [hunk 4]
```diff
@@ -111,6 +120,50 @@ func (s *testLeaderChangeSuite) TestLeaderChange(c *C) {
...
+		for {
+			select {
+			case <-quit:
+				return
+			default:
+			}
+
+			physical, logical, err1 := cli.GetTS()
+			if err1 == nil {
+				c.Assert(lastPhysical<<18+lastLogical, Less, physical<<18+logical)
```
Comment (overvenus) [inline pd-client/leader_change_test.go]: Please add some explanation for the magic number 18.

**Comment by nolouch** [comment]
LGTM

File: pd-client/leader_change_test.go [hunk 5]
```diff
@@ -111,6 +120,50 @@ func (s *testLeaderChangeSuite) TestLeaderChange(c *C) {
...
+
+	etcdCli, err := clientv3.New(clientv3.Config{
+		Endpoints:   endpoints,
+		DialTimeout: time.Second,
+	})
+	c.Assert(err, IsNil)
+	leaderPath := filepath.Join("/pd", strconv.FormatUint(cli.GetClusterID(), 10), "leader")
+	for i := 0; i < 10; i++ {
+		mustWaitLeader(c, servers)
+		_, err = etcdCli.Delete(context.TODO(), leaderPath)
```
Comment (overvenus) [inline pd-client/leader_change_test.go]: What happens when someone updates leader by accident?

Comment (disksing) [inline pd-client/leader_change_test.go]: It won't matter. It's just a hack way to force a new election.

**Pull request merged**